File: src/main/java/ru/femirion/leetcode/medium/AlternatingGroupsII.java
package ru.femirion.leetcode.medium;

/**
 * 3208. Alternating Groups II
 * https://leetcode.com/problems/alternating-groups-ii/description/
 */
public class AlternatingGroupsII {

    /**
     * Runtime 3 ms Beats 86.87%
     * Memory 63.17 MB Beats 50.00%
     */
    public int numberOfAlternatingGroups(int[] colors, int k) {
        int start = 0;
        int end = 1;

        int result = 0;
        int prevColor = colors[0];
        for (int i = 1; i < colors.length + k - 1; i++) {
            int idx = i < colors.length ? i : i - colors.length;
            int curColor = colors[idx];
            if (curColor == prevColor) {
                start = end;
                end++;
            } else {
                if (end - start < k) {
                    end++;
                }
                if (end - start == k) {
                    result++;
                    end++;
                    start++;
                }

            }
            prevColor = curColor;
        }

        return result;
    }
}

File: src/main/go/easy/alternating_groups_II.go
package easy

/**

3208. Alternating Groups II
https://leetcode.com/problems/alternating-groups-ii/description/

Runtime 108 ms Beats 33.33%
Memory 11.70 MB Beats 20.00%
*/

func numberOfAlternatingGroups(colors []int, k int) int {
	start := 0
	end := 1
	result := 0
	prevColor := colors[0]
	for i := 1; i < len(colors)+k-1; i++ {
		var idx int
		if i < len(colors) {
			idx = i
		} else {
			idx = i - len(colors)
		}
		curColor := colors[idx]
		if prevColor == curColor {
			start = end
			end++
		} else {
			if end-start < k {
				end++
			}
			if end-start == k {
				result++
				end++
				start++
			}
		}
		prevColor = curColor
	}
	return result
}
